### ISSUE TITLE
SVG not maintaining aspect ration

### ISSUE THREAD

**Issue by qsc-jhndnn** [SVGKit/SVGKit#522]
Using this SVG ( https://upload.wikimedia.org/wikipedia/commons/c/c0/RATED_PG-13.svg ) I am having issues maintaining aspect ratio when using SVGKFastImageView. I have tried setting contentMode = UIViewContentModeScaleAspectFit and scaleToFitInside: and neither ( or both ) have any affect. In the debugger, svgView._internalSizeThatWasSetExplicitlyByUser does have the correct aspect ratio. Is there anything else I need to do so my image is drawn correctly?

**Comment by adamgit** [comment]
SVG has its own definition of scaling, and it's managed by the image, not
the renderer (this is a better way of doing it, it makes more sense). The
imageview's contentmode will always be ignored.

I'm guessing there's something wrong in SVGKFastImageView that's trying to
auto-scale the SVG to fill the space - there's not much code in that class,
perhaps have a closer look in the debugger?

**Comment by drewcovi** [comment]
I'm a bit confused. I'm also running into this. as explained here, it appears the default is to stretch SVGs not respect the viewbox...

https://useyourloaf.com/blog/stretching-redrawing-and-positioning-with-contentmode/

are we certain there's nothing that can be done to fix this issue in svgkit?

**Comment by adamgit** [comment]
OP never replied with more detail.

Obviously... it works correctly for most people most of the time. One of the images in the demo app is from the SVG spec authors and tests many combinations of stretch, fill, aspect ratio. Last time I looked, svgkit was displaying all of them correctly! 

However, if that image has stopped working then we have a major regression bug. It's thr one with smiley faces in many boxes in different combinations.

If that image is still correct then... This bug report needs more detail: what is wrong and why? 
The link you pasted is unrelated to SVGs. Im not sure how it affects this report.

**Comment by drewcovi** [comment]
@adamgit as per the docs, it appears that SVGKFastImageView is a version of UIImageView which the article references.

http://cocoadocs.org/docsets/SVGKit/1.2.0/Classes/SVGKFastImageView.html

As per that article

```
The default if you do not set a content mode is .scaleToFill shown below left.
.scaleToFill stretches the view to fill the available space without
maintaining the aspect ratio.
```

This reflects the behavior I'm seeing with SVGKFastImageView as implemented in nativescript-svg. By default it appears to discard preserveAspectRatio attributes and defaults to stretching to fill.

**Comment by drewcovi** [comment]
@adamgit if the demo is here, then it appears the few instances of svgkfastimageview never implement aspectratio flags and the content modes are simply set to the default (scaletofill)

https://github.com/SVGKit/SVGKit/blob/0d4db53890c664fb8605666e6fbccd14912ff821/SVGKitLibrary/Demo-iOS/SB%7Eiphone.storyboard

**Comment by adamgit** [comment]
When this was written (has anything changed? Has UIView gained new options 
to support all missing cases?) Apple's UIView does not support as many 
sizing options as SVG requires (including: SVGs can specify their own 
sizing! Apple doesn't let images do that). Therefore I don't see how you 
could support the UIView method unless you disable core features of SVG?

That's an option: we could offer a flag that breaks your SVG rendering 
(some files would be unaffected, others would be affected negatively,  and 
it would be your responsibility to make sure you only ever draw files that 
are unaffected) so that you can use the small subset of sizes Apple 
recognizes. But it would have to default to disabled I think (it breaks the 
SVG spec, by disabling SVGs own builtin scaling behaviour).


TL;DR: there is no perfect solution here, because Apple does not allow 
people to extend their API, and their API does not have enough options to 
cover all rendering situations. They designed it for doing simple bitmap 
only, eg PNG.

**Comment by drewcovi** [comment]
@adamgit I'll dig deeper but as far as I can tell, in most third parties implementing SVGKFastImageView, they're often manually recreating aspect ratio containers and placing the svgkfastimageview within that container (a react plugin is doing this for example).

Perhaps thats the best way forward for now, it just seems that for svgkfastimageview to be of any real value, it shouldn't stretchtofill, and should (as you mention) adhere to the various scaling behaviors.

**Comment by SMH-7** [comment]
Fast forward to 2021.
Upon choosing content mode of scale to fill, it scales my content inside of SVG then SVG. Am stuck...
